core/trino-main/src/main/java/io/trino/operator/window/matcher/ThreadEquivalence.java

```diff
@@ -0,0 +1,235 @@
+ * (`positionsToCompare`). For all those navigating operations, check if they
+ * return the same result (i.e. navigate to the same input row) for both threads.
+ * NOTE: the computations can be simplified by stripping the physical offsets
+ * and skipping navigations which refer to the universal pattern variable.
+ * <p>
+ * 3. for all those labels, get all navigating operations for `CLASSIFIER` calls
+ * which need to be performed during label evaluation (`labelsToCompare`).
+ * For all those navigating operations, check if they navigate to rows tagged
+ * with the same label (but not necessarily the same position in input) for both threads.
+ * <p>
+ * NOTE: the navigating operations for `MATCH_NUMBER` calls can be skipped
+ * altogether, since the match number is constant in this context.
+ */
+public class ThreadEquivalence
+{
+    // for every pointer (instruction) in the program, the set of labels reachable by the program
+    // starting from this instruction until the program ends, through any path.
+    private final List<Set<Integer>> reachableLabels;
+
+    // for every label, the set of navigations for accessing input values based on the defining condition
+    private final List<Set<LogicalIndexNavigation>> positionsToCompare;
+
+    // for every label, the set of navigations for accessing the matched labels based on the defining condition
+    private final List<Set<LogicalIndexNavigation>> labelsToCompare;
+
+    public ThreadEquivalence(Program program, List<List<PhysicalValuePointer>> valuePointers)
+    {
+        this.reachableLabels = computeReachableLabels(program);
+
+        this.positionsToCompare = getInputValuePointers(valuePointers).stream()
+                .map(pointersList -> pointersList.stream()
+                        .map(PhysicalValuePointer::getLogicalIndexNavigation)
+                        .filter(navigation -> !navigation.getLabels().isEmpty())
+                        .map(LogicalIndexNavigation::withoutPhysicalOffset)
+                        .map(ThreadEquivalence::allPositionsToCompare)
+                        .flatMap(Collection::stream)
+                        .collect(toImmutableSet()))
+                .collect(toImmutableList());
+
+        this.labelsToCompare = getClassifierValuePointers(valuePointers).stream()
+                .map(pointersList -> pointersList.stream()
+                        .map(PhysicalValuePointer::getLogicalIndexNavigation)
+                        .map(ThreadEquivalence::allPositionsToCompare)
+                        .flatMap(Collection::stream)
+                        .collect(toImmutableSet()))
+                .collect(toImmutableList());
+    }
+
+    public boolean equivalent(int firstThread, ArrayView firstLabels, int secondThread, ArrayView secondLabels, int pointer)
+    {
+        checkArgument(firstLabels.length() == secondLabels.length(), "matched labels for compared threads differ in length");
```

Comment:
Can't this happen in the case of an alternating pattern with different numbers of labels in each branch?

Reply:
The matching algorithm iterates over the input. At each step of processing, all the existing threads have ingested the same number of input rows, which means that they assigned the same number of labels. The equivalence check is performed for threads within a single step of iteration.